Homework7/Hw7/MyHtmlServices/HtmlHelperExtensions.cs

```diff
@@ -0,0 +1,90 @@
+using System.ComponentModel.DataAnnotations;
+using System.Reflection;
+using System.Text;
+using System.Text.RegularExpressions;
+using Hw7.Models;
+using Microsoft.AspNetCore.Html;
+using Microsoft.AspNetCore.Mvc.Rendering;
+
+namespace Hw7.MyHtmlServices;
+
+public static class HtmlHelperExtensions
+{
+    public static IHtmlContent MyEditorForModel(this IHtmlHelper helper)
+    {
+        TagBuilder builder = new TagBuilder("div");
+        IHtmlContent br = (new TagBuilder("br")).RenderStartTag();
+        var model = helper.ViewData.Model;
+        foreach (var property in helper.ViewData.ModelMetadata.ModelType.GetProperties())
+        {
+            TagBuilder div = new TagBuilder("div");
+            string name = GetDisplayName(property);
+            GetLabelData(property, name, div, br);
+            GetInputData(property, name, div, br);
+            if(model != null)
+                ValidateData(property, model, div, br);
+            builder.InnerHtml.AppendHtml(div);
+        }
+        return builder;
+    }
+
+    private static string GetDisplayName(PropertyInfo propertyInfo)
+    {
+        var attribute = propertyInfo.GetCustomAttribute<DisplayAttribute>();
+        return attribute != null ? attribute.Name! : ParseName(propertyInfo);
```

Comment:
`return attribute?.Name ?? ParseName(propertyInfo)`

Reply:
[?](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators?redirectedfrom=MSDN#null-conditional-operators--and-)

[??](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-coalescing-operator?redirectedfrom=MSDN)